csv-generator/deployment/src/main/java/io/quarkiverse/operatorsdk/csv/deployment/builders/CsvManifestsBuilder.java

```diff
@@ -0,0 +1,328 @@
+            if (iconAsStream != null) {
+                final byte[] iconAsBase64 = Base64.getEncoder()
+                        .encode(iconAsStream.readAllBytes());
+                csvSpecBuilder.addNewIcon()
+                        .withBase64data(new String(iconAsBase64))
+                        .withMediatype("image/png")
+                        .endIcon();
+            }
+        } catch (IOException e) {
+            // ignore
+        }
+
+        String defaultServiceAccountName = NO_SERVICE_ACCOUNT;
+        if (!serviceAccounts.isEmpty()) {
+            defaultServiceAccountName = serviceAccounts.get(0).getMetadata().getName();
+        }
+
+        final var installSpec = csvSpecBuilder.editOrNewInstall()
+                .withStrategy(DEPLOYMENT).editOrNewSpec();
+        handleClusterPermissions(clusterRoleBindings, clusterRoles, roles, defaultServiceAccountName, installSpec);
+        handlePermissions(clusterRoles, roleBindings, roles, defaultServiceAccountName, installSpec);
+        handleDeployments(deployments, installSpec);
+
+        // do not forget to end the elements!!
+        installSpec.endSpec().endInstall();
+        csvSpecBuilder.endSpec();
+
+        final var csv = csvBuilder.build();
+        return YAML_MAPPER.writeValueAsBytes(csv);
+    }
+
+    private String getIconName() {
+        return csvGroupName + ".icon.png";
+    }
+
+    private void handleDeployments(List<Deployment> deployments,
+            NamedInstallStrategyFluent.SpecNested<ClusterServiceVersionSpecFluent.InstallNested<ClusterServiceVersionFluent.SpecNested<ClusterServiceVersionBuilder>>> installSpec) {
+        deployments.forEach(deployment -> handleDeployment(deployment, installSpec));
+    }
+
+    private void handlePermissions(List<ClusterRole> clusterRoles, List<RoleBinding> roleBindings, List<Role> roles,
+            String defaultServiceAccountName,
+            NamedInstallStrategyFluent.SpecNested<ClusterServiceVersionSpecFluent.InstallNested<ClusterServiceVersionFluent.SpecNested<ClusterServiceVersionBuilder>>> installSpec) {
+        Map<String, List<PolicyRule>> customPermissionRules = new HashMap<>();
+        if (metadata.permissionRules != null) {
+            for (CSVMetadataHolder.PermissionRule permissionRule : metadata.permissionRules) {
+                String serviceAccountName = StringUtils.defaultIfEmpty(permissionRule.serviceAccountName,
+                        defaultServiceAccountName);
+                List<PolicyRule> customRulesByServiceAccount = customPermissionRules.get(serviceAccountName);
+                if (customRulesByServiceAccount == null) {
+                    customRulesByServiceAccount = new LinkedList<>();
+                    customPermissionRules.put(serviceAccountName, customRulesByServiceAccount);
+                }
+
+                List<String> verbs;
+                if (permissionRule.verbs != null && permissionRule.verbs.length > 0) {
```

Comment:
The defaults should be encoded in the annotation itself so that the `CSVMetadataHolder` contains all the information when built in one spot instead of distributing the logic across the code, if possible…

Reply:
Done

Reply:
This logic should be put in a getter for `verbs`, imo.